apps/speed.c

```diff
@@ -3337,6 +3458,178 @@ int speed_main(int argc, char **argv)
         }
     }
 
+# ifndef OPENSSL_NO_SM2
+    for (testnum = 0; testnum < SM2_NUM; testnum++) {
+        int st = 1;
+        EVP_PKEY *sm2_pkey = NULL;
+        EVP_PKEY_CTX *pctx = NULL;
+        EVP_PKEY_CTX *sm2_pctx = NULL;
+        EVP_PKEY_CTX *sm2_vfy_pctx = NULL;
+        size_t sm2_sigsize = 0;
+
+        if (!sm2_doit[testnum])
+            continue;           /* Ignore Curve */
+        /* Init signing and verification */
+        for (i = 0; i < loopargs_len; i++) {
+            loopargs[i].sm2_ctx[testnum] = EVP_MD_CTX_new();
+            if (loopargs[i].sm2_ctx[testnum] == NULL) {
+                st = 0;
+                break;
+            }
+            loopargs[i].sm2_vfy_ctx[testnum] = EVP_MD_CTX_new();
+            if (loopargs[i].sm2_vfy_ctx[testnum] == NULL) {
+                st = 0;
+                break;
+            }
+
+            /* SM2 keys are generated as normal EC keys with a special curve */
+            if ((pctx = EVP_PKEY_CTX_new_id(EVP_PKEY_EC, NULL)) == NULL
+                || EVP_PKEY_keygen_init(pctx) <= 0
+                || EVP_PKEY_CTX_set_ec_paramgen_curve_nid(pctx,
+                    test_sm2_curves[testnum].nid) <= 0
+                || EVP_PKEY_keygen(pctx, &sm2_pkey) <= 0) {
+                st = 0;
+                EVP_PKEY_CTX_free(pctx);
+                break;
+            }
+            /* free previous one and alloc a new one */
+            EVP_PKEY_CTX_free(pctx);
+
+#ifndef FIPS_MODE
```

Comment:
Since the `ECDSA_size` is defined within the non-`FIPS_MODE` in the library, is there any where that an application can check if the function is available or not?

Reply:
Ah, you are misunderstanding the meaning of `FIPS_MODE` I think. Various code files now get compiled *twice*. The FIPS module does not link against libcrypto and has its own copy of a lot of the code. The `FIPS_MODE` macro is defined if the code is being compiled for inclusion in the FIPS module for its own *internal* use. It is never defined when compiling it for inclusion in libcrypto. When an application calls `ECDSA_size()` it is always calling the libcrypto version - so an application does not care whether it is available in the FIPS module or not.

Reply:
Ahh, now I see the point.